<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>Decombinator by innate2adaptive</title>
    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/github-dark.css">
    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.7.1/jquery.min.js"></script>
    <script src="javascripts/respond.js"></script>
    <!--[if lt IE 9]>
      <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <!--[if lt IE 8]>
    <link rel="stylesheet" href="stylesheets/ie.css">
    <![endif]-->
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

  </head>
  <body>
      <div id="header">
        <nav>
          <li class="fork"><a href="https://github.com/innate2adaptive/Decombinator">View On GitHub</a></li>
          <li class="downloads"><a href="https://github.com/innate2adaptive/Decombinator/zipball/master">ZIP</a></li>
          <li class="downloads"><a href="https://github.com/innate2adaptive/Decombinator/tarball/master">TAR</a></li>
          <li class="title">DOWNLOADS</li>
        </nav>
      </div><!-- end header -->

    <div class="wrapper">

      <section>
        <div id="title">
          <h1>Decombinator</h1>
          <p>Decombinator v3: fast, error-correcting analysis of TCR repertoires</p>
          <hr>
          <span class="credits left">Project maintained by <a href="https://github.com/innate2adaptive">innate2adaptive</a></span>
          <span class="credits right">Hosted on GitHub Pages &mdash; Theme by <a href="https://twitter.com/michigangraham">mattgraham</a></span>
        </div>

        <h1>
<a id="innate2adaptive--decombinator" class="anchor" href="#innate2adaptive--decombinator" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>innate2adaptive / Decombinator</h1>

<h2>
<a id="v31" class="anchor" href="#v31" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>v3.1</h2>

<h5>
<a id="this-version-written-by-james-m-heather-niclas-thomas-katharine-best-theres-oakes-mazlina-ismail-and-benny-chain" class="anchor" href="#this-version-written-by-james-m-heather-niclas-thomas-katharine-best-theres-oakes-mazlina-ismail-and-benny-chain" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>This version written by James M. Heather, Niclas Thomas, Katharine Best, Theres Oakes, Mazlina Ismail and Benny Chain</h5>

<h5>
<a id="innate2adaptive-lab--university-college-london-2016" class="anchor" href="#innate2adaptive-lab--university-college-london-2016" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Innate2Adaptive lab @ University College London, 2016</h5>

<hr>

<p>Decombinator is a fast and efficient tool for the analysis of T-cell receptor (TCR) repertoire sequences produced by deep sequencing. The current version (v3) improves upon the previous through:</p>

<ul>
<li>incorporation of error-correction strategies using unique molecular index (UMI) barcoding techniques</li>
<li>optimisation for additional speed and greater accuracy</li>
<li>usage of extended tag sets of V and J genes allowing detection of genes of all functionalities</li>
<li>greater command line input, log file production and code commenting, providing greater flexibility, repeatability and ease of modification</li>
</ul>

<h2>
<a id="for-the-original-decombinator-paper-please-see-thomas-et-al-bioinformatics-2013" class="anchor" href="#for-the-original-decombinator-paper-please-see-thomas-et-al-bioinformatics-2013" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>For the original Decombinator paper, please see <a href="http://dx.doi.org/10.1093/bioinformatics/btt004">Thomas <em>et al</em>, Bioinformatics, 2013</a>.</h2>

<h3 id="top">
<a id="navigation" class="anchor" href="#navigation" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Navigation:</h3>

<ul>
<li><a href="#installation">Installation</a></li>
<li><a href="#demultiplexing">Demultiplexing</a></li>
<li><a href="#decombinator">Decombining</a></li>
<li><a href="#collapsing">Collapsing</a></li>
<li><a href="#cdr3">CDR3 extraction</a></li>
<li><a href="#generalusage">General usage notes</a></li>
<li><a href="#calldcr">Calling Decombinator from other scripts</a></li>
</ul>

<hr>

<p>TCR repertoire sequencing (TCRseq) offers a powerful means to investigate biological samples to see what sequences are present in what distribution. However current high-throughput sequencing (HTS) technologies can produce large amounts of raw data, which presents a computational burden to analysis. This such raw data will also unavoidably contain errors relative to the original input molecules, which could confound the results of any experiment.</p>

<p>Decombinator addresses the problem of speed by employing a rapid and <a href="https://figshare.com/articles/Aho_Corasick_String_Matching_Video/771968">highly efficient string matching algorithm</a> to search the FASTQ files produced by HTS machines for rearranged TCR sequence. The central algorithm searches for 'tag' sequences, the presence of which uniquely indicates the inclusion of particular V or J genes in a recombination. If V and J tags are found, Decombinator can then deduce where the ends of the germline V and J gene sections are (i.e. how much nucleotide removal occurred during V(D)J recombination), and what nucleotide sequence (the 'insert sequence') remains between the two. These five pieces of information - the V and J genes used, how many deletions each had and the insert sequene - contain all of the information required to reconstruct the whole TCR nucleotide sequence, in a more readily stored and analysed way. Decombinator therefore rapidly searches through FASTQ files and outputs these five fields into comma delimited output files, one five-part classifier per line.</p>

<p>The Decombinator suite of scripts are written in <strong>Python v2.7</strong>, and the default parameters are set to analyse data as produced by the ligation-mediated 5' RACE TCR amplification pipeline. The pipeline consists of four scripts, which are applied sequentially to the output of the previous starting with TCR-containing FASTQ files (produced using the barcoding 5' RACE protocol):</p>

<ol>
<li>Raw FASTQ files are demultiplexed to individual samples</li>
<li>Sample specific files are then searched for rearranged TCRs with Decombinator</li>
<li>Decombined data is error-corrected ('collapsed') with reference to the random barcode sequences added prior to amplification</li>
<li>Error-corrected data is translated and CDR3 sequences are extracted</li>
</ol>

<hr>

<h1 id="installation">
<a id="installation-and-getting-started" class="anchor" href="#installation-and-getting-started" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Installation and getting started</h1>

<p><em>Please note that some familiarity with Python and command-line script usage is assumed, but everything should be easily managed by those without strong bioinformatic or computational expertise.</em></p>

<h3>
<a id="required-modules" class="anchor" href="#required-modules" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Required modules</h3>

<p>Python 2.7 is required to run this pipeline, along with the following non-standard modules:</p>

<ul>
<li>acora</li>
<li>biopython</li>
<li>regex</li>
<li>python-Levenshtein</li>
</ul>

<p>These can be installed via pip (although most will likely appear in other package managers), e.g.:</p>

<div class="highlight highlight-source-shell"><pre>pip install python-biopython python-levenshtein regex acora</pre></div>

<p>If users are unable to install Python and the required modules, a Python installation complete with necessary packages has been bundled into a Docker container, which should be runnable on most setups. The appropriate image is located <a href="https://hub.docker.com/r/decombinator/dcrpython/">on Dockerhub, under the name 'dcrpython'</a>.</p>

<h3>
<a id="get-scripts" class="anchor" href="#get-scripts" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Get scripts</h3>

<p>The Decombinator scripts are all downloadable from the <a href="https://github.com/innate2adaptive/Decombinator">innate2adaptive/Decombinator Github repository</a>. These can be easily downloaded in a Unix-style terminal like so:</p>

<div class="highlight highlight-source-shell"><pre>git clone https://github.com/innate2adaptive/Decombinator.git</pre></div>

<p>Decombinator also requires a number of additional files, which contain information regarding the V and J gene sequences, their tags, and the locations and motifs which define their CDR3 regions. By default Decombinator downloads these files from <a href="https://github.com/innate2adaptive/Decombinator-Tags-FASTAs">the git repository where they are maintained</a>, which obviously requires a working internet connection. In order to run Decombinator offline, these files must be downloaded to a local location, and either stored within the directory where you wish to run Decombinator or specified using the appropriate command line flag. These additional files can also be downloaded via <code>git clone</code>:</p>

<div class="highlight highlight-source-shell"><pre>git clone https://github.com/innate2adaptive/Decombinator-Tags-FASTAs.git</pre></div>

<p>The current version of Decombinator has tag sets for the analysis of alpha/beta and gamma/beta TCR repertoires from both mouse and man. In addition to the original tag set developed for the 2013 Thomas <em>et al</em> Bioinformatics paper, an 'extended' tag set has been developed for human alpha/beta repertoires, which covers 'non-functional' V and J genes (ORFs and pseudogenes, according to IMGT nomenclature) as well as just the functional.</p>

<h3>
<a id="general-notes" class="anchor" href="#general-notes" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>General notes</h3>

<ul>
<li>All scripts accept gzipped or uncompressed files as input</li>
<li>By default, all output files will be gz compressed

<ul>
<li>This can be suppressed by using the "don't zip" flag: <code>-dz</code>
</li>
</ul>
</li>
<li>All files also output a summary log file

<ul>
<li>We strongly recommend you familarise yourself with them.</li>
</ul>
</li>
<li>All options for a given script can be accessed by using the help flag <code>-h</code>
</li>
</ul>

<p><sub><a href="#top">↑Top</a></sub></p>

<hr>

<h1 id="demultiplexing">
<a id="demultiplexorpy" class="anchor" href="#demultiplexorpy" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Demultiplexor.py</h1>

<h2>
<a id="demultiplexing-getting-sample-specific-barcoded-vdj-containing-reads" class="anchor" href="#demultiplexing-getting-sample-specific-barcoded-vdj-containing-reads" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Demultiplexing: getting sample-specific, barcoded V(D)J containing reads</h2>

<p><em>This demultiplexing step is designed specifically to make use of the random barcode sequences introduced during the Chain lab's wet lab TCR amplification protocol. While it should be fairly straightforward to adapt this to other UMI-based approaches, that will require some light modification of the scripts. Users wanting to apply Decombinator to demultiplexed, non-barcoded data can skip this step.</em></p>

<p>Illumina machines produce 3 FASTQ read files:</p>

<ul>
<li>Read 1 

<ul>
<li>Contains the V(D)J sequence and a demultiplexing index (the SP1 index)</li>
</ul>
</li>
<li>Read 2 

<ul>
<li>Contains the barcode sequences, and reads into the 5' UTR</li>
</ul>
</li>
<li>Index 1 

<ul>
<li>Contains the second demultiplexing index (the typical Illumina index read, giving the SP2 index)</li>
</ul>
</li>
</ul>

<p>The demultiplexing script extracts the relevant sequences from each read file and combines them into reads in one file, and demultiplexes them on the two indexes to produce separate sample files containing both the V(D)J and barcode information.</p>

<p>Note that you may need to run <code>bcl2fastq</code> or alter a configuration file on your sequencing machine in order to obtain the I1 index read file (which is not advised unless you know what you're doing, please consult whoever runs your machines). For example, this can be achieved on a MiSeq by adding the following line to the <code>Reporter.exe.config</code>, under <code>&lt;appsettings&gt;</code>:</p>

<div class="highlight highlight-source-shell"><pre><span class="pl-k">&lt;</span>add key=”CreateFastqForIndexReads” value=”1”/<span class="pl-k">&gt;</span></pre></div>

<p>You need to provide the demultiplexing script:</p>

<ul>
<li>The location of the 3 read files</li>
<li>An index csv file giving sample names and indexes

<ul>
<li>Sample names will be carried downstream, so use sensible identifiers </li>
<li>Including the chain (e.g. 'alpha') will allow auto-detection in subsequent scripts (if <em>only one</em> chain is used per file)</li>
<li>Do not use space or '.' characters</li>
</ul>
</li>
<li>Give numeric indexes, SP1 end first, then SP2 end

<ul>
<li>e.g. <code>'AlphaSample1,5,9'</code> would put all reads amplified using SP1-6N-I-5-aRC1 and P7-L-9 in one file</li>
</ul>
</li>
</ul>

<p>An example command might look like this:</p>

<div class="highlight highlight-source-shell"><pre>python Demultiplexor.py -r1 R1.fq.gz -r2 R2.fq.gz -i1 I1.fq.gz -ix IndexFile.ndx</pre></div>

<p>It's also likely that your read files are already demultiplexed by the machine, using just the SP2 indexes alone. Single read files can be produced using bash to output all appropriate reads into one file, e.g.:</p>

<div class="highlight highlight-source-shell"><pre><span class="pl-c"># Linux machines</span>
zcat <span class="pl-cce">\*</span>R1<span class="pl-cce">\*</span> <span class="pl-k">|</span> gzip <span class="pl-k">&gt;</span> R1.fq.gz
<span class="pl-c"># Macs</span>
gunzip -c <span class="pl-cce">\*</span>R1<span class="pl-cce">\*</span> <span class="pl-k">|</span> gzip <span class="pl-k">&gt;</span> R1.fq.gz</pre></div>

<p>If you suspect your indexing might be wrong you can use the <code>'outputall'</code> flag (<code>-a</code>). This outputs demultiplexed files using all possible combination of indexes used in our protocol, and can be used to identify the actual index combinations contained in a run and locate possible cross-contamination. Note that this is only recommended for troubleshooting and not for standard use as it will decrease the number of successfully demultiplexed reads per samples due to fuzzy index matching. </p>

<p>Addition of new index sequences will currently require some slight modification of the code, although if people requested the use of a more easily edited external index file that could be incorporated in the next update.</p>

<p><sub><a href="#top">↑Top</a></sub></p>

<hr>

<h1 id="decombinator">
<a id="decombinatorpy" class="anchor" href="#decombinatorpy" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Decombinator.py</h1>

<h2>
<a id="decombining-identifying-rearranged-tcrs-and-outputting-their-5-part-classifiers" class="anchor" href="#decombining-identifying-rearranged-tcrs-and-outputting-their-5-part-classifiers" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Decombining: identifying rearranged TCRs and outputting their 5-part classifiers</h2>

<p>This script performs the key functions of the pipeline, as it searches through demultiplexed reads for rearranged TCR sequences. It looks for short 'tag' sequences (using Aho-Corasick string matching): the presence of a tag uniquely identifies a particular V or J gene. If it finds both a V and a J tag (and the read passes various filters), it assigns the read as recombined, and outputs a five-part Decombinator index (or 'DCR'), which uniquely represents a given TCR rearrangement.</p>

<p>All DCR-containing output files are comma-delimited, with the fields of that five part classifier containing, in order:</p>

<ul>
<li>The V index (which V gene was used)</li>
<li>The J index</li>
<li>Number of V deletions (relative to germ line)</li>
<li>Number of J deletions</li>
<li>Insert sequence (the nucleotide sequence between end of deleted V and J)</li>
</ul>

<p>The V and J indices are arbitrary numbers based on the order of the tag sequences in the relevant tag file (using Python indexing, which starts at 0 rather than 1). Also note that the number of V and J deletions actually just represents how many bases have been removed from the end of that particular germline gene (as given in the germline FASTA files in the additional file repo); it is entirely possible that more bases where actually deleted, and just that the same bases have been re-added. </p>

<p>Various additional fields may follow the five part classifier, but the DCR will always occupy the first five positions. An example identifier, from a human alpha chain file, might look like this:</p>

<div class="highlight highlight-source-shell"><pre>1, 22, 9, 0, CTCTA</pre></div>

<p>Which corresponds to a rearrangement between TRAV1-2 (V index <strong>1</strong>, with <strong>9</strong> nucleotides deleted) and TRAJ33 (J index <strong>22</strong>, with <strong>0</strong> deletions), with an insert sequence (i.e. non-templated additions to the V and/or the J gene) of '<strong>CTCTA</strong>'. For beta chains, the insert sequence will contain any residual TRBD nucleotides, although as these genes are very short, homologous and typically highly 'nibbled', they are often impossible to differentiate.</p>

<p><code>Decombinator.py</code> needs to be provided with a FASTQ file, using the <code>-fq</code> flag (which, if following the Chain lab protocol will have come from the output of <code>Demultiplexor.py</code>). Users can also specify:</p>

<ul>
<li>The species: <code>-sp</code>

<ul>
<li>'human' (default) or 'mouse'</li>
</ul>
</li>
<li>The TCR chain locus: <code>-c</code>

<ul>
<li>Just use the first letter, i.e. a/b/g/d</li>
<li>Not required if unambiguously specified in filename</li>
<li>This flag will override automatic chain detection</li>
</ul>
</li>
</ul>

<div class="highlight highlight-source-shell"><pre>python Decombinator.py -fq AlphaSample1.fq.gz -sp mouse -c a</pre></div>

<p>The other important flag is that which specifies the DNA strand orientations that are searched for rearrangements: <code>-or</code>, for which users can specify 'forward', 'reverse' or 'both'. As the 5' RACE approach used in the Chain lab means that all recombinations will be read from the constant region, the default is set to reverse.</p>

<p>Decombinator also outputs a number of additional fields by default, in order to facilitate error-correction in subsequent processes. These additional fields are, in order:</p>

<ul>
<li>The FASTQ ID (i.e. the first line of four, following the '@' character)

<ul>
<li>This allows users to go back to the original FASTQ file and find whole reads of interest</li>
</ul>
</li>
<li>The 'inter-tag' sequence, and quality

<ul>
<li>These two fields are the nucleotide sequence and Q scores of the original FASTQ, running from the start of the found V tag to the end of the found J tag </li>
<li>This therefore just lifts the part of the sequence that contains the rearrangement/CDR3 sequences, which will be used for error-correction</li>
</ul>
</li>
<li>The barcode sequence, and quality

<ul>
<li>The first 30 bases of each read, carried over from R2 during demultiplexing, contains the two random nucleotide sequences which make up the barcode (UMI)</li>
</ul>
</li>
</ul>

<div class="highlight highlight-source-shell"><pre>43, 19, 5, 2, TCGACCTC, M01996:15:000000000-AJF5P:1:1101:17407:1648, AAGTGTCAGACTCAGCGGTGTACTTCTGTGCTCTGTCGACCTCAACAGAGATGACAAGATCATCTTTGGAAAAGGGACACG, HHHGHFHHHHG<span class="pl-k">?</span>FGECHHHHHHHHHGHHFHHGGGGFFHHGFGGHHHHHHHHHHHHHHHHHHHHGHHHHHHHHHHHHHGGGH, GTCGTGATCGGCCGGTCGTGATCGTGCACA, 1AA@A<span class="pl-k">?</span>FFF<span class="pl-k">??</span>DFCGGGEG<span class="pl-k">?</span>FGGAGHGFHC</pre></div>

<p>By default Decombinator assigns a 'dcr_' prefix and '.n12' file extension to the output files, to aid in downstream command line data handling. However these can can be altered using the <code>-pf</code> and <code>-ex</code> flags respectively.</p>

<p>As Decombinator uses tags to identify different V/J genes it can only detect those genes that went into the tag set. Both human and mouse have an 'original' tag set, which contains all of the prototypical alleles for each 'functional' TCR gene (as defined by IMGT).
There is also an 'extended' tag set for human alpha/beta genes, which includes tags for the prototypical alleles of all genes, regardless of predicted functionality. This is the default and recommended tag set for humans (due to increased specificity and sensitivity), but users can change this using the <code>-tg</code> flag.</p>

<p>Users wanting to run <code>Decombinator.py</code> on their own, non-barcoded data should set the flag <code>-nbc</code>. Choosing this option will stop Decombinator from outputting the additional fields required for demultiplexing: instead each classifier will simply be appended an abundance, indicating how many times each identifier appeared in that run. Note that this changes the default file extension to '.nbc', which users may wish to change.</p>

<p>The supplementary files required by Decombinator can be downloaded by the script from GitHub as it runs. If running offline, you need to <a href="https://github.com/innate2adaptive/Decombinator-Tags-FASTAs">download the files</a> to the directory where you run Decombinator, or specifically inform it of the path using the <code>-tfdir</code> flag.</p>

<p>Supplementary files are named by the following convention:
[<em>species</em>]_[<em>tag set</em>]_[<em>gene type</em>].[<em>file type</em>]<br>
E.g.:
<code>human_extended_TRAV.tags</code></p>

<p>Also please note the TCR gene naming convention, as determined by IMGT:<br>
TRAV1-2*01<br>
TR = TCR / A = alpha / V = variable / 1 = family / -2 = subfamily / *01 = allele</p>

<p><sub><a href="#top">↑Top</a></sub></p>

<hr>

<h1 id="collapsing">
<a id="collapsinatorpy" class="anchor" href="#collapsinatorpy" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Collapsinator.py</h1>

<h2>
<a id="collapsing-using-the-random-barcodes-to-error--and-frequency-correct-the-repertoire" class="anchor" href="#collapsing-using-the-random-barcodes-to-error--and-frequency-correct-the-repertoire" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Collapsing: using the random barcodes to error- and frequency-correct the repertoire</h2>

<p>Assuming that Decombinator has been run on data produced using the Chain lab's (or a comparable) UMI protocol, the next step in the analysis will remove errors and duplicates produced during the amplification and sequencing reactions.</p>

<p>The barcode sequence is contained in one of the additional fields output by <code>Decombinator.py</code> in the .n12 files, that which contains the first 30 bases of R2. As Illumina sequencing is particularly error-prone in the reverse read, and that reads can be phased (i.e. they do not always begin with the next nucleotide that follows the sequencing primer) our protocol uses known spacer sequences to border the random barcode bases, so that we can identify the actual random bases. The hexameric barcode locations (N6) are determined in reference to the two spacer sequences like so:</p>

<pre><code>I8 (spacer) – N6 – I8 – N6 – 2 base overflow (n)
GTCGTGATNNNNNNGTCGTGATNNNNNNnn
</code></pre>

<p>The collapsing script can therefore use the spacer sequences to be sure we have found the right barcode sequences.</p>

<p>The <code>Collapsinator.py</code> script performs the following procedures:</p>

<ul>
<li>Scrolls through each DCR of the input .n12 file</li>
<li>First performs error-correction (removing TCR reads produced by errors from e.g. PCR)

<ul>
<li>Then looks at all the different DCRs associated with a given barcode</li>
<li>Takes the most common inter-tag sequence/DCR combination as the 'true' TCR (as errors are likely to occur during a later PCR cycle, and thus will most often be minority variants, see <a href="http://dx.doi.org/10.1002/eji.201242517">Bolotin <em>et al.</em>, 2012</a>).</li>
<li>Compare this against all other DCRs that shared the same barcode</li>
<li>Those that only have a few differences are likely erroneous branches, and can be discounted</li>
<li>Frequent but significantly different TCRs could represent 'barcode clash', and thus the process repeats</li>
</ul>
</li>
<li>Second the script estimates the true cDNA frequency 

<ul>
<li>Clusters all barcode sequences found in conjunction with a DCR (within a given threshold) </li>
<li>This is required as there could be errors within the barcode sequences themselves</li>
<li>Number of final clusters gives a better estimate of the original frequency of that cDNA molecule</li>
</ul>
</li>
<li>Outputs a DCR identifier plus an additional sixth field, giving the corrected abundance of that TCR in the sample</li>
</ul>

<p>Collapsing occurs in a chain-blind manner, and so only the decombined '.n12' file is required, without any chain designation, with the only required parameter being the infile:</p>

<div class="highlight highlight-source-shell"><pre>python Collapsinator.py -in dcr_AlphaSample1.n12</pre></div>

<p>A number of the filters and thresholds can be altered using different command line flags. In particular, changing the R2 barcode quality score and TCR sequence edit distance thresholds (via the <code>-mq</code> <code>-bm</code> <code>-aq</code> and <code>-lv</code> flags) are the most influential parameters. However this the need for such fine tuning will likely be very protocol-speficic, and is only suggested for advanced users, and with careful data validation.</p>

<p>The default file extension is '.freq'.</p>

<p><sub><a href="#top">↑Top</a></sub></p>

<hr>

<h1 id="cdr3">
<a id="cdr3translatorpy" class="anchor" href="#cdr3translatorpy" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>CDR3translator.py</h1>

<h2>
<a id="tcr-translation-and-cdr3-extraction-turning-dcr-indexes-into-complementarity-determining-region-3-sequences" class="anchor" href="#tcr-translation-and-cdr3-extraction-turning-dcr-indexes-into-complementarity-determining-region-3-sequences" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>TCR translation and CDR3 extraction: turning DCR indexes into complementarity determining region 3 sequences</h2>

<p>As the hypervariable region and the primary site of antigenic contact, the CDR3 is probably the region of most interest. By convention, the <a href="http://dx.doi.org/10.1016/S0145-305X(02)00039-3">CDR3 is defined</a> as running from the position of the second conserved cysteine encoded in the 3' of the V gene to the phenylalanine in the conserved 'FGXG' motif in the J gene. However, some genes use non-canonical residues/motifs, and the position of these motifs varies.</p>

<p>In looking for CDR3s, we also find 'non-productive' reads, i.e. those that don't appear to be able to make productive, working TCRs. This is determined based on the presence of stop codons, being out of frame, or lacking appropriate CDR3 motifs. </p>

<p>The process occurs like so:</p>

<pre><code># Starting with a Decombinator index
43, 5, 1, 7, AGGCAGGGATC

# Used to construct whole nucleotide sequences, using the germline FASTAs as references
GATACTGGAGTCTCCCAGAACCCCAGACACAAGATCACAAAGAGGGGACAGAATGTAACTTTCAGGTGTGATCCAATTTCTGAACACAACCGCCTTTATTGGTACCGACAGACCCTGGGGCAGGGCCCAGAGTTTCTGACTTACTTCCAGAATGAAGCTCAACTAGAAAAATCAAGGCTGCTCAGTGATCGGTTCTCTGCAGAGAGGCCTAAGGGATCTTTCTCCACCTTGGAGATCCAGCGCACAGAGCAGGGGGACTCGGCCATGTATCTCTGTGCCAGCAGCTTAGAGGCAGGGATCAATTCACCCCTCCACTTTGGGAATGGGACCAGGCTCACTGTGACAG

# This is then translated into protein sequence
DTGVSQNPRHKITKRGQNVTFRCDPISEHNRLYWYRQTLGQGPEFLTYFQNEAQLEKSRLLSDRFSAERPKGSFSTLEIQRTEQGDSAMYLCASSLEAGINSPLHFGNGTRLTVT

# The CDR3 sequence is then extracted based on the conserved C and FGXG motifs (as stored in the .translate supplementary files)
CASSLEAGINSPLHF
</code></pre>

<p>In order to do so, a third kind of supplementary data file is used, .translate files, which provide the additional information required for CDR3 extraction for each gene type (a/b/g/d, V/J). They meet the same naming conventions as the tag and FASTA files, and consist of four comma-delimited fields, detailing:</p>

<ul>
<li>Gene name</li>
<li>Conserved motif position (whether C or FGXG)</li>
<li>Conserved motif sequence (to account for the the non-canonical)</li>
<li>IMGT-defined gene functionality (F/ORF/P)</li>
</ul>

<p><code>CDR3translator.py</code> only requires an input file, and to be told which TCR chain is being examined (although this can also be inferred from the file name, as with <code>Decombinator.py</code>. Input can be collapsed or not (e.g. .freq or .n12), and the program will retain the frequency information if present.</p>

<div class="highlight highlight-source-shell"><pre>python CDR3translator.py -in dcr_AlphaSample1.freq
<span class="pl-c"># or</span>
python CDR3translator.py -in dcr_AnotherSample1.freq -c b</pre></div>

<p>Outputs a '.cdr3' file by default, which looks like:
<code>CASSISGRLDTQYF, 1</code></p>

<p>Alternatively you can use the 'dcr output' flag (<code>-do</code>) to output DCRs alongside the CDR3s they encode, e.g.:
<code>18, 8, 5, 5, ATCAGCGGGAGATT:CASSISGRLDTQYF, 1</code></p>

<p>You can also use the 'nonproductive' flag  (<code>-np</code>) to output non-productive rearrangements to a separate file:
<code>15, 3, 1, 5, CTGTATCAGGGGGCC:OOF_with_stop, 1</code></p>

<p>The residues that make up the 'GXG' section of the 'FGXG' motif can also be included if desired, through use of the <code>-gxg</code> flag.</p>

<p><sub><a href="#top">↑Top</a></sub></p>

<hr>

<h1 id="generalusage">
<a id="general-usage-notes-and-tips" class="anchor" href="#general-usage-notes-and-tips" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>General usage notes and tips</h1>

<h2>
<a id="being-efficient" class="anchor" href="#being-efficient" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Being efficient</h2>

<ul>
<li>Use the help flag (<code>-h</code>) to see all options for a given script</li>
<li>Where possible, run scripts in same folder as data

<ul>
<li>Should work remotely, but recommended not to</li>
</ul>
</li>
<li>Use bash loops to iterate over many samples

<ul>
<li>For example, to run the CDR3 translation script on all alpha .freq files in the current directory:
<code>for x in *alpha*.freq.gz; do echo $x; python ligTCRtranslateCDR3.py -in $x -c a; done</code>
</li>
</ul>
</li>
<li>There is also a <a href="https://github.com/innate2adaptive/automate-decombinator">makefile in development</a> to automate the pipeline from start to finish</li>
<li>Ideally don't pool alpha and beta of one individual into same index for sequencing

<ul>
<li>You may wish to know the fraction of reads Decombining, which will be confounded if there are multiple samples or chains per index combination</li>
</ul>
</li>
</ul>

<h2>
<a id="be-aware-of-the-defaults" class="anchor" href="#be-aware-of-the-defaults" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Be aware of the defaults</h2>

<ul>
<li>These scripts are all set up to analyse data produced from our ligation TCR protocol, and thus the defaults reflect this

<ul>
<li>By far the most common libraries analysed to date are human αβ repertoires</li>
</ul>
</li>
<li>If running on mouse and/or gamma/delta repertoires, it may be worth altering the defaults towards the top of the code to save on having to repeatedly have to set the defaults</li>
<li>If running on data not produced using our protocol, likely only the Decombinator and translation scripts will be useful 

<ul>
<li>The <code>-nbc</code> (non-barcoding) flag will likely need to be set (unless you perform analogous UMI-positioning to <code>Demultiplexor.py</code>)</li>
<li>If your libraries contain TCRs in the forward orientation, or in both, this will also need to set with the <code>-or</code> flag</li>
</ul>
</li>
</ul>

<h2>
<a id="keep-track-of-whats-going-on" class="anchor" href="#keep-track-of-whats-going-on" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Keep track of what's going on</h2>

<ul>
<li>Using the 'dontgzip' flag ('-dz') will stop the scripts automatically compressing the output data, which will minimise processing time (but take more space)

<ul>
<li>It is recommend to keep gzipped options as default</li>
<li>Compressing/decompressing data slightly slows your analysis, but dramatically reduces the data storage footprint</li>
</ul>
</li>
<li>Don't be afraid to look at the code inside the scripts to figure out what's going on</li>
<li>Check your log files, as they contain important information</li>
<li>
<strong>Back up your data</strong> (including your log files)!</li>
</ul>

<h2>
<a id="use-good-naming-conventions" class="anchor" href="#use-good-naming-conventions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Use good naming conventions</h2>

<ul>
<li>Keep separate alpha and beta files, named appropriately

<ul>
<li>Keep chain type in name</li>
<li>Decombinator will always take specified chain as the actual, but failing that will look in filename</li>
</ul>
</li>
<li>Don't use full stops in file names

<ul>
<li>Some of the scripts use them to determine file extension locations</li>
<li>Additionally discourages inadvertent data changing</li>
</ul>
</li>
<li>While there are default file extensions, these settings can be set via the <code>-ex</code> flag 

<ul>
<li>Similarly, when Decombining you can set the prefix with <code>-pf</code>
</li>
</ul>
</li>
</ul>

<h2>
<a id="using-docker" class="anchor" href="#using-docker" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Using Docker</h2>

<ul>
<li>Docker may be convenient for systems where installing many different packages is problematic

<ul>
<li>It should allow users to run Decombinator scripts without installing Python + packages (although you still have to install Docker)</li>
<li>The scripts however will run faster if you natively install Python on your OS</li>
</ul>
</li>
<li>If opted for, users must first <a href="https://docs.docker.com/engine/installation/">install Docker</a>
</li>
<li>Then scripts can be run like so:</li>
</ul>

<div class="highlight highlight-source-shell"><pre>docker run -it --rm --name dcr -v <span class="pl-s"><span class="pl-pds">"</span><span class="pl-smi">$PWD</span><span class="pl-pds">"</span></span>:/usr/src/myapp -w /usr/src/myapp decombinator/dcrpython python SCRIPT.py<span class="pl-s"><span class="pl-pds">`</span></span></pre></div>

<ul>
<li>Depending on your system, you may need to run this as a superuser, i.e. begin the command <code>sudo docker...</code>
</li>
<li>N.B.: the first time this is run Docker will download <a href="https://hub.docker.com/r/decombinator/dcrpython/">the Decombinator image from Dockerhub</a>, which will make it take longer than usual</li>
</ul>

<p><sup><a href="#top">↑Top</a></sup></p>

<hr>

<h1 id="calldcr">
<a id="calling-decombinator-from-other-python-scripts" class="anchor" href="#calling-decombinator-from-other-python-scripts" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Calling Decombinator from other Python scripts</h1>

<p>As an open source Python script, <code>Decombinator</code> (and associated functions) can be called by other scripts, allowing advanced users to incorporate our rapid TCR assignation into their own scripts, permitting the development of highly bespoke functionalities.</p>

<p>If you would like to call Decombinator from other scripts you need to make sure you fulfil the correct requirements upstream of actually looking for rearrangements. You will thus need to:</p>

<ul>
<li>Set up some 'dummy' input arguments, as there are some that Decombinator will expect</li>
<li>Run the 'import_tcr_info' function, which sets up your environment with all of the required variables (by reading in the tags and FASTQs, building the tries etc)</li>
</ul>

<p>An example script which calls Decombinator might therefore look like this:</p>

<div class="highlight highlight-source-python"><pre><span class="pl-k">import</span> Decombinator <span class="pl-k">as</span> dcr

<span class="pl-c"># Set up dummy command line arguments arguments </span>
inputargs <span class="pl-k">=</span> {<span class="pl-s"><span class="pl-pds">"</span>chain<span class="pl-pds">"</span></span>:<span class="pl-s"><span class="pl-pds">"</span>b<span class="pl-pds">"</span></span>, <span class="pl-s"><span class="pl-pds">"</span>species<span class="pl-pds">"</span></span>:<span class="pl-s"><span class="pl-pds">"</span>human<span class="pl-pds">"</span></span>, <span class="pl-s"><span class="pl-pds">"</span>tags<span class="pl-pds">"</span></span>:<span class="pl-s"><span class="pl-pds">"</span>extended<span class="pl-pds">"</span></span>, <span class="pl-s"><span class="pl-pds">"</span>tagfastadir<span class="pl-pds">"</span></span>:<span class="pl-s"><span class="pl-pds">"</span>no<span class="pl-pds">"</span></span>, <span class="pl-s"><span class="pl-pds">"</span>lenthreshold<span class="pl-pds">"</span></span>:<span class="pl-c1">130</span>, <span class="pl-s"><span class="pl-pds">"</span>allowNs<span class="pl-pds">"</span></span>:<span class="pl-c1">False</span>, <span class="pl-s"><span class="pl-pds">"</span>fastq<span class="pl-pds">"</span></span>:<span class="pl-s"><span class="pl-pds">"</span>blank<span class="pl-pds">"</span></span>}

<span class="pl-c"># Initalise variables</span>
dcr.import_tcr_info(inputargs)

testseq <span class="pl-k">=</span> <span class="pl-s"><span class="pl-pds">"</span>CACTCTGAAGATCCAGCGCACACAGCAGGAGGACTCCGCCGTGTATCTCTGTGCCAGCAGCTTATTAGTGTTAGCGAGCTCCTACAATGAGCAGTTCTTCGGGCCAGGGACACGGCTCACCGTGCTAGAGGACCTGAAA<span class="pl-pds">"</span></span>

<span class="pl-c"># Try and Decombine this test sequence</span>
<span class="pl-c1">print</span> dcr.dcr(testseq, inputargs)</pre></div>

<p>This script produces the following list output:
<code>[42, 6, 2, 0, 'TTAGTGTTAGCGAG', 22, 118]</code></p>

<p>The first five fields of this list correspond to the standard DCR index as described above, while the last two indicate the start position of the V tag and the end position of the J tag (thus definining the 'inter-tag region').</p>

<p>Users may wish to therefore loop through their data that requires Decombining, and call this function on each iteration. However bear in mind that run in this manner Decombinator only checks whatever DNA orientation it's given: if you need to check both orientations then I suggest running one orientation through first, checking whether there's output, and if not then running the reverse orientation through again (which can be easily obtained using the Seq functionality of the Biopython package).</p>

<p><sup><a href="#top">↑Top</a></sup></p>

<hr>

<h3>
<a id="legacy-decombinator-versions" class="anchor" href="#legacy-decombinator-versions" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Legacy Decombinator versions</h3>

<p>If users wish to view previous versions of Decombinator, v2.2 is available from the old GitHub repo <a href="https://github.com/uclinfectionimmunity/Decombinator/">uclinfectionimmunity / Decombinator</a>. However it is not recommended that this script be used for analysis as it lacks some key error-reduction features that have been integrated into subsequent versions, and is no longer supported.</p>

<hr>

<h3>
<a id="related-reading" class="anchor" href="#related-reading" aria-hidden="true"><span aria-hidden="true" class="octicon octicon-link"></span></a>Related reading</h3>

<ul>
<li>
<a href="http://dx.doi.org/10.1093/bioinformatics/btt004">Thomas et al (2013), Bioinformatics, “Decombinator: a tool for fast, efficient gene assignment in T-cell receptor sequences using a finite state machine”</a>

<ul>
<li>The original Decombinator paper, which explains the principles of the script </li>
</ul>
</li>
<li>
<a href="http://dx.doi.org/10.3389/fimmu.2015.00644">Heather et al (2016), Frontiers in Immunology, “Dynamic Perturbations of the T-Cell Receptor Repertoire in Chronic HIV Infection and following Antiretroviral Therapy”</a>

<ul>
<li>Paper describing the previous version of the protocol and some applications (as applied to studying the repertoire of HIV+ patients), which also introduces the collapsing procedure</li>
</ul>
</li>
<li>
<a href="http://dx.doi.org/10.1038/srep14629">Best et al (2015), Scientific Reports, “Computational analysis of stochastic heterogeneity in PCR amplification efficiency revealed by single molecule barcoding”</a>

<ul>
<li>An in-depth analysis of barcoded data showing the need for barcoding in producing robustly quantitative data from the stochastic maelstrom of amplification that makes up PCR</li>
</ul>
</li>
</ul>
      </section>

    </div>
    <!--[if !IE]><script>fixScale(document);</script><![endif]-->
    
  </body>
</html>
